### PR TITLE
Major simplification of webpage.sh + extend DHCP to IPv6

### LINKED ISSUE #1034: Nest thermostat offline after installing Pi-Hole
**In raising this issue, I confirm the following (please check boxes, eg [X]) Failure to fill the template will close your issue:**

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [x] The issue I am reporting can be *replicated*
- [x] The issue I'm reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the codebase?:**

_1_

---
**[FEATURE REQUEST | QUESTION | OTHER]:**

Please [submit your feature request here](https://discourse.pi-hole.net/c/feature-requests), so it is votable by the community.  It's also easier for us to track.

**[BUG | ISSUE] Expected Behaviour:**
Nest thermostat should be online and should be controlled by Nest mobile app/web UI.  I also have Nest cameras and they are working fine.

**[BUG | ISSUE] Actual Behaviour:**
Nest thermostat shows offline and not cannot be accessed from Mobile app/web UI.

**[BUG | ISSUE] Steps to reproduce:**
Install Nest thermostat.  Install Pi-Hole connected to the router.
-
-
-
-

**(Optional) Debug token generated by `pihole -d`:**
1rn87fe1wd

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following (please check boxes, eg [X]) _Failure to fill the template will close your PR_:**

***Please submit all pull requests against the `development` branch. Failure to do so will delay or deny your request***

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/pi-hole/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:** 10

---
Internal PR. Will make it much simpler to add additional settings in the future.

Will also add IPv6 DHCP functionality.

Since DNS settings will be reprocessed after an update and DHCP settings will be stored in a different file, this will fix #1034 

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

### CODE REVIEW

**Comment by darkrain-nl** [comment]
Is it possible/a good idea to make the type of DHCPv6 configurable? Now it is always DHCPv6 while Android does not support that, for Android SLAAC is needed. So it would be nice if we could choose between SLAAC and DHCPv6...

My configuration for SLAAC looks like this b.t.w.:

`# Do stateless DHCP, SLAAC, and generate DNS names for SLAAC addresses
# from DHCPv4 leases.
dhcp-range=2001:here my prefix::, ra-stateless, ra-names
ra-param=*,0,0
dhcp-option=option6:dns-server,[::]`

This works on Android, Debian Jessie and Windows. One thing is not working without a 3rd party solution and that's RDNSS on Windows. But that is no big problem as ipv6 DNS request can go via ipv4 DNS servers...

**Comment by DL6ER** [comment]
@darkrain-nl I'm completely open for suggestions concerning our implementation of IPv6.
@dschaper might also want to comment on this.

**Comment by dschaper** [comment]
screenshot from 2016-12-28 10-03-03

**Comment by dschaper** [comment]
Although generating a SLAAC from the IPv4 (privacy extentions) would be a better approach than MAC'ing it, but I think that's set at the client stack level and not at the DHCP level. LEDE(OpenWRT) allows for a hint in hexadecimal which is the last `n` bits of the host portion of the address. So I can add a hint of `0000 000A` and that's either used directly as the LSB of the host, or masked on to the LSBs of the host.

**Comment by darkrain-nl** [comment]
@dschaper yes, you are correct about the privacy extensions. That is a client level configuration, Windows enables privacy extensions by default. 
I'm not sure if you can configure dnsmasq to be stateless + stateful? That would be great, because then you don't have to make a choice anymore...

**Comment by dschaper** [comment]
I'll have to pull out the configs and see what comes up, I think odhcpd is used on the router solutions, but if there's a way to do it in `dnsmasq` we should be able to find it. (The documentation is quite extensive actually.)

**Comment by darkrain-nl** [comment]
If you can find it, it just means I wasn't searching good engough ;-)
I did search for this some months ago...

**Comment by dschaper** [comment]
Night time reading material http://www.thekelleys.org.uk/dnsmasq/docs/dnsmasq-man.html

**Comment by darkrain-nl** [comment]
@dschaper thanks I was aware of the manual.

**Comment by dschaper** [comment]
That was a post to remind me of where I needed to start looking, and I'll check that WP site as well. Thanks!

**Comment by darkrain-nl** [comment]
Ok I did some testing and found out it is possible to do stateful DHCPv6 and SLAAC.

My new configuration:
`dhcp-range=::2,::500,constructor:eth0,ra-names,slaac

ra-param=*,0,0
dhcp-option=option6:dns-server,[::]`

This works on both Android and Windows. No need for any 3rd party applications anymore.

**Comment by DL6ER** [comment]
If you find something that will make all of us happy and both of you agree that this is good, then I'll just put it in. The less settings we expose the our users the better their experience. And I have to do less programming on the Settings page ;-)

**Comment by dschaper** [comment]
If it does both stateful and stateless, then bases are covered, and the clients can battle it out for addressing.

**Comment by DL6ER** [comment]
Is there something where we can be sure that they catch the correct IPv6 DNS server? On my Ubuntu system at home, sometimes the automatic IPv6 config finds the router (although it shouldn't do anything concerning DHCP, etc.) and sometimes the Pi.

**Comment by dschaper** [comment]
I think that's more in the RA portion. (Admittedly I haven't gone to deep into router advertisements and payloads...) But `option6` is the DNS server field for DHCP so even if they go SLAAC, they should be able to get that information.

**Comment by darkrain-nl** [comment]
@DL6ER how is your router configured? In mine it is possible to just do SLAAC + stateless DHCPv6 which means it only advertises the default gateway for the network and it handles the SLAAC portion (Android devices in my case) It doesn't send the option6 field.
Also what you're describing is a problem only for Linux based systems, Windows doesn't have that problem cause Windows does nothing with DNS servers from Router Advertising.

**Comment by DL6ER** [comment]
Well, my router is the cheap one I got for free. No settings at all concerning IPv6. The only thing I can change is the IPv4 DHCP server (which is disabled).

**Comment by darkrain-nl** [comment]
That's a shame, cause what happens now is that both dnsmasq and your router do Router Advertisements. This is somewhat a good thing, because you need the Router Advertisement from the router for your default gateway. But on the other hand you don't want it to do something with the DNS server which should be coming from dnsmasq.
I'm not sure how to prevent this from happening other then buying a new router ;-)

**Comment by DL6ER** [comment]
Are we good to go now with the current version of this PR?

**Comment by dschaper** [comment]
We'll have to see if the others can help test this out. I have a 5 vLAN configuration that I'd rather not break down quite yet. I'll see if I can set up something smaller and sniff a few packets.

**Comment by darkrain-nl** [comment]
I'm not sure if I look correct, but is the IPv4 DNS server address (the pi-hole) also distributed by DHCP?
Something like `dhcp-option=6,0.0.0.0` 

Another thing in the future could be the option to create static leases based on MAC address or hostname, but that can be something for later..

**Comment by DL6ER** [comment]
> Another thing in the future could be the option to create static leases based on MAC address or hostname, but that can be something for later..

We almost have this. You can simply edit the `dhcp.leases` file.

> I'm not sure if I look correct, but is the IPv4 DNS server address (the pi-hole) also distributed by DHCP?
Something like dhcp-option=6,0.0.0.0

The DHCP server of `dnsmasq` defaults to itself as DNS server. Use the IPv4 part already with great success at home and I know of many others where it works just fine.

**Comment by darkrain-nl** [comment]
> The DHCP server of dnsmasq defaults to itself as DNS server. Use the IPv4 part already with great success at home and I know of many others where it works just fine.

I did not know this.
I did know I can just edit the dhcp.leases file, or add some dhcp-host lines in the dhcp config file.

Looks good from my side.

**Comment by darkrain-nl** [comment]
One more question, why use an infinite lease time?
I think it's better to use something like 12 or 24 hours, if a device leaves the network I think it's nice it cleans up after some time...

**Comment by DL6ER** [comment]
I can only speak for myself but I like to know that the IPs in my network don't change. Imagine I'm away for a week and switch on my devices afterwards. Would like them to be reachable under the same address. I doubt that you would normally see enough devices in a net to outrun the /24 range

**Comment by dschaper** [comment]
My vote is for user configurable. I normally run 7 day leases, but if I know there's going to be some changes in the infrastructure, being able to step down the lease time to a short half-life is nice, then I can change things out and set the lease back up to a high level. I also have guests on net that I don't need hanging around so I don't need them to be perma-leases. I'd rather have my servers on static/reserved lease and then the transient devices on a timed lease. I don't know if a client would attempt to rebind if given an infinite lease, there's no real TTL for it to calculate a time period to send out a rebind request?

**Comment by darkrain-nl** [comment]
> I can only speak for myself but I like to know that the IPs in my network don't change. Imagine I'm away for a week and switch on my devices afterwards. Would like them to be reachable under the same address.
That's what DNS is for in my opinion.

> I don't know if a client would attempt to rebind if given an infinite lease, there's no real TTL for it to calculate a time period to send out a rebind request?
I know the DHCP server sends out infinite as lease time, but I don't know how a client behaves with this.

My vote is also for user configurable.

File: advanced/Scripts/webpage.sh [hunk 1]
```diff
@@ -149,123 +173,136 @@ RestartDNS(){
...
+dhcp-authoritative
+dhcp-range=${DHCP_START},${DHCP_END},infinite
+dhcp-option=option:router,${DHCP_ROUTER}
+dhcp-leasefile=/etc/pihole/dhcp.leases
+domain=${PIHOLE_DOMAIN}
+quiet-dhcp
+quiet-dhcp6
+#enable-ra
+dhcp-option=option6:dns-server,[::]
+dhcp-range=::100,::1ff,constructor:${interface},ra-names,slaac
```
Comment (darkrain-nl) [inline advanced/Scripts/webpage.sh]: Can you please add the lease time for dhcpv6 also? It seems to default to 1 hour which is a bit fast in my opinion...
Maybe make it configurable for both dhcpv4 and dhcpv6.

**Comment by DL6ER** [comment]
From the `dnsmasq` man page:

> When replying with information from /etc/hosts or configuration or the DHCP leases file dnsmasq by default sets the time-to-live field to zero, meaning that the requester should not itself cache the information. This is the correct thing to do in almost all situations.

File: advanced/Scripts/webpage.sh (hunk 1 above)
Comment (DL6ER) [inline advanced/Scripts/webpage.sh]: From the `dnsmasq` man page:
> When replying with information from /etc/hosts or configuration or the DHCP leases file dnsmasq by default sets the time-to-live field to zero, meaning that the requester should not itself cache the information. This is the correct thing to do in almost all situations.

**Comment by darkrain-nl** [comment]
OK, I see now that it doesn't make a difference in relation to how often a client requests a rebind.
So the default should be sufficient.

**Comment by DL6ER** [comment]
Okay, I added the lease time to the settings page (see PR https://github.com/pi-hole/AdminLTE/pull/309)

I managed to apply it for IPv4 + IPv6:
```
Dec 29 14:24:51 dnsmasq-dhcp[32081]: DHCP, IP range 10.1.0.2 -- 10.1.9.2, lease time 1d
Dec 29 14:24:51 dnsmasq-dhcp[32081]: DHCPv6, IP range ::100 -- ::1ff, lease time 1d, template for eth0
Dec 29 14:24:51 dnsmasq-dhcp[32081]: DHCPv4-derived IPv6 names on eth0
Dec 29 14:24:51 dnsmasq-dhcp[32081]: router advertisement on eth0
```

**Comment by DL6ER** [comment]
Enabled the DHCP logging by default. Not sure if we want to do this, but we can discuss. I don't want to add another setting for this on the Settings page...

**Comment by darkrain-nl** [comment]
> Enabled the DHCP logging by default. Not sure if we want to do this, but we can discuss. I don't want to add another setting for this on the Settings page...

I like it as you know, so my vote goes to enable.

**Comment by darkrain-nl** [comment]
I just found out I had no IPv6 DNS server address configured on my Windows 10 machine, at first I thought that there was a problem with dnsmasq, but it seems Microsoft broke DHCPv6 after their Anniversary Update... https://social.technet.microsoft.com/Forums/windows/en-US/b16e7d78-e390-4ada-a24b-3ccba60fa571/no-ipv6-dns-statelessdhcp-since-windows-10-anniversary-update?forum=win10itpronetworking&prof=required

I'm posting this just so you know.

**Comment by PromoFaux** [comment]
Approved

**Pull request merged**